apps/shipments/views.py

```diff
@@ -99,24 +99,63 @@ def get_tracking_data(self, request, version, pk):
 
         if tracking_data.exists():
             response = Template('{"data": $geojson}').substitute(geojson=render_point_features(shipment, tracking_data))
-            return HttpResponse(content=response,
-                                content_type='application/vnd.api+json')
+            httpresponse = HttpResponse(content=response, content_type='application/vnd.api+json')
+            return httpresponse
 
         return HttpResponseNotFound()
 
+    @method_decorator(cache_page(60 * 10, cache='page'))  # Cache responses for 10 minutes
+    @action(detail=True, methods=['get'], permission_classes=(IsOwnerOrShared,))
+    def tracking(self, request, version, pk):
+        return self.get_tracking_data(request, version, pk)
+
+    def update(self, request, *args, **kwargs):
+        """
+        Update the shipment with new details, overwriting the built-in method
+        """
+        partial = kwargs.pop('partial', False)
+        instance = self.get_object()
+        LOG.debug(f'Updating shipment {instance.id} with new details.')
+        log_metric('transmission.info', tags={'method': 'shipments.update', 'module': __name__})
+
+        serializer = ShipmentUpdateSerializer(instance, data=request.data, partial=partial,
+                                              context={'auth': get_jwt_from_request(request)})
+        serializer.is_valid(raise_exception=True)
+
+        shipment = self.perform_update(serializer)
+        async_jobs = shipment.asyncjob_set.filter(state__in=[JobState.PENDING, JobState.RUNNING])
+        response = ShipmentTxSerializer(shipment)
+
+        response.instance.async_job_id = async_jobs.latest('created_at').id if async_jobs else None
+
+        return Response(response.data, status=status.HTTP_202_ACCEPTED)
+
+
+class DeviceViewSet(viewsets.GenericViewSet):
+    queryset = Device.objects.all()
+    permission_classes = permissions.AllowAny
+    serializer_class = PermissionLinkSerializer
+
     def add_tracking_data(self, request, version, pk):
-        shipment = Shipment.objects.get(pk=pk)
+        LOG.debug(f'Adding tracking data by the device.')
+        log_metric('transmission.info', tags={'method': 'devices.tracking', 'module': __name__})
+        device = Device.objects.get(pk=pk)
+        try:
+            shipment = Shipment.objects.filter(device=device).first()
+        except ObjectDoesNotExist:
```

Comment:
This will never be thrown by first():
https://docs.djangoproject.com/en/2.1/ref/models/querysets/#first

You are mixing fetching approaches here. `Shipment.objects.get(device_id=pk)` returns exceptions (MultipleObjectsReturned, ObjectDoesNotExist). First() returns either the first value, or None.
https://docs.djangoproject.com/en/2.1/ref/models/querysets/#get